2021q2/kernel-sanitizers.adoc

```diff
@@ -0,0 +1,31 @@
+=== Kernel Sanitizers
+
+Contact: Mark Johnston <markj@FreeBSD.org>
+
+Sanitizers are a bug-detection facility which use a combination of instrumentation inserted by the compiler (LLVM in this case) and runtime state tracking to detect bugs in C code.
+They can automatically detect many types of C programming bugs, such as use-after-frees and uses of uninitialized variables, which may otherwise require substantial effort to identify.
+They are particularly effective in combination with regression testing suites or fuzzing tools such as link:https://github.com/google/syzkaller[syzkaller].
+Unlike tools such as Valgrind, software must be recompiled to enable a given sanitizer, but sanitizers can be used in the kernel.
+Kernels with sanitizers enabled incur a significant performance overhead from the runtime, in both CPU utilization and memory usage.
+
+Work has been ongoing to port a pair of kernel sanitizers from NetBSD to FreeBSD.
+These are the Kernel Address SANitizer (KASAN) and Kernel Memory SANitizer (KMSAN).
+The KASAN port is complete and has been committed to FreeBSD's development branch, and a small LLVM patch to enable KASAN and KMSAN on FreeBSD has also been committed.
+KMSAN intercepts all memory accesses and verifies that they are valid using some hidden state saved for each memory allocation in the kernel; see man:kasan[9] for further details.
+It can be enabled in amd64 kernels by compiling the GENERIC-KASAN kernel configuration.
+The FreeBSD regression test suite currently passes with KASAN enabled.
+
+The KMSAN port is currently in progress and is nearing completion.
+KMSAN detects uses of uninitialized memory using the algorithms described in link:https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43308.pdf[the original MemorySanitizer paper].
+In particular, it can detect instances of uninitialized kernel memory being copied out to userspace.
+Kernel subsystems may additionally call into the KMSAN runtime to verify the state of a given buffer.
+This can be used to, for example, add checks in the network stack for uninitialized bytes in egress packets.
```

Comment:
I'd move "to" after "for example". (I don't mind split infinitives, but a comma-separated embedded phrase in-between feels too much for my taste.)